### DISCUSSION TITLE
[Q&A] 自定义插件出现调用报错怎么办？

### DISCUSSION THREAD

**Discussion by Payne-Wang** [lobehub/lobehub#876] [General Question] [answered]
### 💻 系统环境 | Operating System

Windows

### 🌐 浏览器 | Browser

Chrome

### 🐛 问题描述 | Bug Description

使用Chatgpt plus支持的一些插件的时候，总会出现各种不匹配和报错。
比如这个scholar AI，昨晚用的好好的，今天就出现问题了。
另外请问这跟部署方式有关吗？lobechat我是部署在vercle上的，没用docker部署过
谢谢大佬

报错如下:
```json
{
  "error": {
    "status": 400,
    "statusCode": 400,
    "response": {
      "ok": false,
      "url": "https://plugin.scholar-ai.net/api/abstracts?keywords=HMGCS2&sort=date&query=HMGCS2&peer_reviewed_only=true&start_year=2021",
      "status": 400,
      "statusText": "Bad Request",
      "headers": {
        "access-control-allow-origin": "http://localhost:3000",
        "alt-svc": "h3=\":443\"; ma=2592000",
        "apx-hit": "true",
        "content-length": "100",
        "content-type": "application/json",
        "date": [
          "Fri",
          "29 Dec 2023 03:43:14 GMT"
        ],
        "server": [
          "Caddy",
          "Google Frontend"
        ],
        "vary": "Origin",
        "x-cloud-trace-context": "0bd51863e8a77ed967e94ac28854ca10"
      },
      "text": "{\n  \"message\": \"Invalid sort parameter. Only valid values are cited_by_count or publication_date\"\n}\n",
      "data": "{\n  \"message\": \"Invalid sort parameter. Only valid values are cited_by_count or publication_date\"\n}\n",
      "body": {
        "message": "Invalid sort parameter. Only valid values are cited_by_count or publication_date"
      },
      "obj": {
        "message": "Invalid sort parameter. Only valid values are cited_by_count or publication_date"
      }
    }
  }
}
```

[a comment by an automated account is not shown]

**Comment by arvinxx** [comment]
没有关系。虽然 LobeChat 的插件机制上兼容了 OpenAI 的 ChatGPT 插件，但是是否可用我们没法保证。

**Reply by Payne-Wang** [reply]
是的，很奇怪的是装上插件有时候可以用，有时候却会出现参数形式调用错误的问题。Chatgpt plus是用了什么特殊的prompt，使传参形式更严格了吗

**Reply by arvinxx** [reply]
不是，应该是你使用的插件服务方可能对调用的传参有要求，不符合规范就报错。而 gpt 生成的 fc 调用参数有时候是会出问题，导致不符合插件服务端的要求规范，进而报错。例如现在内置的 Dalle 3 也是会有类似问题： https://github.com/lobehub/lobe-chat/issues/812

**Reply by arvinxx** [reply, marked as answer]
要求完整符合 FC 的入参规范是对模型本身有要求的。根据最近的一些反馈来看，哪怕是 GPT-4 在 Function Call 上也仍然存在不遵循指令的情况。这个是LobeChat作为应用层没法解决的问题。只能等未来 gpt-4.5 或者 gpt-5 的模型能力优化了。

**Reply by Payne-Wang** [reply]
明白了，谢谢，很棒的回答！

**Reply by Payne-Wang** [reply]
突然想到，每次插件报错的时候是否可以加一层prompt，可以将报错信息自动发送给GPT，而不是自动返回报错，输出的信息是GPT对报错的参数进行优化并对用户提出改进建议？

**Reply by arvinxx** [reply]
感觉可以试试。我回头构想一下这块可能要怎么做

[a comment by an automated account is not shown]

**Reply by ccdanpian** [reply]
@Payne-Wang 请问如何您是通过什么方法使用的这个自定义插件啊，是下载到本地服务器，还是直接用的官方地址啊？
我也想尝试用一下其他的插件。

**Reply by arvinxx** [reply]
看下文档 https://github.com/lobehub/lobe-chat/blob/main/docs/Usage/Plugins.zh-CN.md#%E5%AE%89%E8%A3%85%E8%87%AA%E5%AE%9A%E4%B9%89%E6%8F%92%E4%BB%B6

**Reply by ccdanpian** [reply]
不知道去哪里找这些第三方插件的地址，有没有一个插件市场啊？

**Comment by ccdanpian** [comment]
找到一个网站，里面收集了不少插件，https://chatbotsplace.com/plugins。

不过还有个问题，比如说试用了另外一个可以读github的插件，Talk with Code: https://talkwithcode.say-apps.com/.well-known/openapi.yaml。

需要授权登录：
1、在chatgpt里返回结果如下， 复制链接后，可以点击去授权
{
  "auth_url": "https://github.com/login/oauth/authorize?client_id=Iv1.93b7a578f760fd72&scope=repo&state=fa00fbbf-9f28-598b-96f5-25012c095a6e",
  "exc": true,
  "m": "We can't find this file or repository. It's might be unavailable for us or it's not exists. Try to authorize and give access to Github Application from the auth_url",
  "useful_urls": {
    "website_url": "https://say-apps.com"
  }
}
2、在lobechat里插件返回结果如下，用这个链接进不到授权页面。
{
  "auth_url": "https://github.com/login/oauth/authorize?client_id=Iv1.93b7a578f760fd72&scope=repo&state=None",
  "exc": true,
  "m": "We can't find this file or repository. It's might be unavailable for us or it's not exists. Try to authorize and give access to Github Application from the auth_url",
  "useful_urls": {
    "website_url": "https://say-apps.com"
  }
}

**Reply by arvinxx** [reply]
https://www.futurepedia.io/ai-plugins/ 这个是我开发时候用的。里面有不少可以参考的


当时测试的记录 https://github.com/lobehub/lobe-chat/discussions/585#discussioncomment-7862195

**Reply by ccdanpian** [reply]
今天在chatgpt网页使用AI PDF插件的时候，发现一个新情况，AI PDF插件停用了，改成了GPTs。并且和插件一样，都是在他的服务器上传pdf，然后把地址复制过来，这是不是说明GPTs看起来也可以使用API？

**Reply by arvinxx** [reply]
是的 gpts 新增了一个actions 的东西，填写一些openapi 的schema 就可以让gpts来调用这个 api。

其实和现在 lobechat 的助手模式是差不多的。你可以设计一个助手，并给他开启相应的插件。这样这个助手就成为了你的某个领域的小帮手。

**Reply by ccdanpian** [reply]
建议lobechat也可以开发（接入）一些后台服务，比如另外一个话题里提到的chatgpt上的github分析插件（总结、代码分析、代码指导等）、还有PDF/网页文章的逐页翻译等等。

现在插件里的AskYourPDF只能实现PDF对话功能，并且好像也不能很准确的进行对话（不太能给出论文中的细节），希望至少能有全文翻译功能。

**Reply by arvinxx** [reply]
有一些场景其实可能用其他工具会更方便，比如论文全文翻译使用沉浸式翻译的阅读体验可能会更佳。

因为 lobe-chat的主模态是 chat，这会使得其在特定交互模态下的效率并不高。

代码理解和分析这个功能我一直想做的，后续会在适合的时候做方案出来
